File: python/120_triangle.py
import sys

class Solution:
    def __init__(self):
        self._min = sys.maxsize
        self.memo = dict()
        self.triangle = None
        
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        self.triangle = triangle
        return self._findMinPath(0,0)
        
    def _findMinPath(self, i, j):
        if i + 1 == len(self.triangle):
            return self.triangle[i][j]
        if (i, j) in self.memo:
            return self.memo[(i, j)]
        tmp = sys.maxsize
        for k in [j, j + 1]:
            tmp = min(tmp, self.triangle[i][j] + self._findMinPath(i + 1, k))
        self.memo[(i,j)] = tmp
        return tmp

File: cpp/120_triangle.cpp
#include <climits>
#include <unordered_map>
#include <utility>
#include <algorithm>

class Solution {
public:
    int min = INT_MAX;
    vector<vector<int>> triangle;
    std::unordered_map<int, int> memo;
    int minimumTotal(vector<vector<int>>& triangle) {
        this->triangle = triangle;
        return this->findMinPath(0,0);
    }
private:
    int findMinPath(int i, int j){
        if(i + 1 == this->triangle.size())
            return this->triangle[i][j];
        if(this->memo.find(cantorPairingHash(i,j)) != this->memo.end())
            return this->memo[cantorPairingHash(i,j)];
        int tmp = INT_MAX;
        for(int k = j; k < j + 2; ++k)
            tmp = std::min(tmp, this->triangle[i][j] + this->findMinPath(i + 1, k));
        this->memo[cantorPairingHash(i,j)] = tmp;
        return tmp;
    }
    float cantorPairingHash(int a, int b){
        return 0.5 * (a + b) * (a + b + 1) + b;
    }
};